sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala

```diff
@@ -169,6 +170,14 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
       }
     }
   }
+
+  test("SPARK-24322 Fix incorrect workaround for bug in java.sql.Timestamp") {
+    withTempPath { path =>
+      val ts = Timestamp.valueOf("1900-05-05 12:34:56.000789")
+      Seq(ts).toDF.write.orc(path.getCanonicalPath)
+      checkAnswer(spark.read.orc(path.getCanonicalPath), Row(ts))
+    }
+  }
```

Comment:
I added the test case for ORC-306 and update the PR title.

Reply:
Does that mean the Hive ORC reader works, but the native ORC reader has the bug?

Reply:
please explicitly set hive reader to native for this test.

Reply:
Oh, I missed this comments. Hive ORC and ORC MR reader doesn't have this bug because it uses `java.sql.Timestamp` class to unserialize it. This happens when we directly access the ORC column's sub-vectors, `times` and `nanos`.

Reply:
`OrcSourceSuite` is dedicated for `native` Orc Reader . For `hive` ORC reader, `HiveOrcSourceSuite`.